test/e2e/pre_stop.go

```diff
@@ -0,0 +1,167 @@
+package e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/wait"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// partially cloned from webserver.go
+type State struct {
+	Received map[string]int
+}
+
+func testPreStop(c *client.Client, ns string) {
+	// This is the server that will receive the preStop notification
+	podDescr := &api.Pod{
+		ObjectMeta: api.ObjectMeta{
+			Name: "server",
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				api.Container{
+					Name:  "server",
+					Image: "gcr.io/google_containers/nettest:1.3",
+					Ports: []api.ContainerPort{{ContainerPort: 8080}},
+				},
+			},
+		},
+	}
+	By(fmt.Sprintf("Creating server pod %s in namespace %s", podDescr.Name, ns))
+	_, err := c.Pods(ns).Create(podDescr)
+	expectNoError(err, fmt.Sprintf("creating pod %s", podDescr.Name))
+
+	// At the end of the test, clean up by removing the pod.
+	defer func() {
+		By("deleting the server pod")
+		c.Pods(ns).Delete(podDescr.Name, nil)
+	}()
+
+	By("Waiting for pods to come up.")
+	err = waitForPodRunningInNamespace(c, podDescr.Name, ns)
+	expectNoError(err, fmt.Sprintf("waiting for server pod to start"))
+
+	val := "{\"Source\": \"prestop\"}"
+
+	podOut, err := c.Pods(ns).Get(podDescr.Name)
+	expectNoError(err, "getting pod info")
+
+	preStopDescr := &api.Pod{
+		ObjectMeta: api.ObjectMeta{
+			Name: "tester",
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				api.Container{
+					Name:    "tester",
+					Image:   "busybox",
+					Command: []string{"sleep", "600"},
+					Lifecycle: &api.Lifecycle{
+						PreStop: &api.Handler{
+							Exec: &api.ExecAction{
+								Command: []string{
+									"wget", "-O-", "--post-data=" + val, fmt.Sprintf("http://%s:8080/write", podOut.Status.PodIP),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	By(fmt.Sprintf("Creating tester pod %s in namespace %s", podDescr.Name, ns))
+	_, err = c.Pods(ns).Create(preStopDescr)
+	expectNoError(err, fmt.Sprintf("creating pod %s", preStopDescr.Name))
+	deletePreStop := true
+
+	// At the end of the test, clean up by removing the pod.
+	defer func() {
+		if deletePreStop {
+			By("deleting the tester pod")
```

Comment:
By clauses should consistently use sentence case (first letter capitalized), as per https://github.com/golang/go/wiki/CodeReviewComments#error-strings